server/src/listing.rs

```diff
@@ -72,6 +73,34 @@ impl PartyFinderListing {
         slots
     }
 
+    pub fn joinable_roles(&self) -> u32 {
+        let one_player_per_job = self
+            .search_area
+            .contains(SearchAreaFlags::ONE_PLAYER_PER_JOB);
+        let mut jobs = JobFlags::empty();
+        let mut jobs_taken = JobFlags::empty();
+        for i in 0..self.slots_available as usize {
+            if i >= self.jobs_present.len() {
+                break;
+            }
+
+            match JOBS.get(&u32::from(self.jobs_present[i])) {
```

Comment:
Realistically, `jobs_present` should always be smaller or equal in size as `slots_available`: so why not iterate over its indices instead?

```suggestion
    for (i, present_job) in self.jobs_present.iter().copied().enumerate() {
            match JOBS.get(&(present_job as u32)) {
```

Because you're also using `i` to index into `slots`, a bounds check on that would be appropriate though (although that edge case would rarely hit, I'd much rather the filter misbehave than the server panic :) )

Reply:
`slots_available` can be smaller than `jobs_present` length (which is always 8 from what I see) if slots are omitted, but yeah, iterating over `jobs_present` should be cleaner